torchtune/utils/_profiler.py

```diff
@@ -48,3 +61,405 @@ def trace_handler(prof) -> None:
+    """
+
+    # Set up profiler activities
+    activities = []
+    if cpu:
+        activities.append(torch.profiler.ProfilerActivity.CPU)
+    if cuda:
+        activities.append(torch.profiler.ProfilerActivity.CUDA)
+    if len(activities) == 0:
+        _warn("No activities specified, defaulting to CPU + CUDA")
+        activities = _DEFAULT_PROFILER_ACTIVITIES
+
+    # Set up profiler schedule
+    use_default_schedule = not any([wait, warmup, active, repeat])
+
+    # Use default schedule if None, else validate that schedule is valid and can be passed to `instantiate`
+    if use_default_schedule:
+        wait = _DEFAULT_SCHEDULE["wait"]
+        warmup = _DEFAULT_SCHEDULE["warmup"]
+        active = _DEFAULT_SCHEDULE["active"]
+        repeat = _DEFAULT_SCHEDULE["repeat"]
+        _warn(
+            " No schedule found in config, defaulting to wait {wait}, warmup {warmup}, active {active}, repeat {repeat}"
+        )
+    else:
+        if not all([wait, warmup, active]):
+            raise ValueError(
+                "Invalid schedule config: must specify wait, warmup, and active"
+            )
+        if repeat is None:
+            _warn(
+                """ No repeat found in schedule config, setting to 1 (one cycle).
+                If you want to cycle continuously, specify `repeat = 0`"""
+            )
+            repeat = 1
+
+    schedule = torch.profiler.schedule(
+        wait=wait, warmup=warmup, active=active, repeat=repeat
+    )
+
+    # profile_memory requires with_stack and record_shapes, hence we override these if profile_memory is True
+    # See torch.profiler.profiler._memory_profile
+    if profile_memory:
+        _warn(
+            "`profile_memory` requires `with_stack` and `record_shapes`, these will be enabled since `profile_memory` is True"
+        )
+    with_stack = with_stack or profile_memory
+    record_shapes = record_shapes or profile_memory
+    # experimental config is needed to export stacks: see https://github.com/pytorch/pytorch/issues/100253
+    experimental_config = _ExperimentalConfig(verbose=True) if with_stack else None
+
+    # Handle exporting of trace, memory timeline and other profiler artifacts
+    if output_dir is None:
+        _warn(
+            f" No output directory found in profiler config, defaulting to {_DEFAULT_PROFILE_DIR}"
+        )
+        output_dir = _DEFAULT_PROFILE_DIR
+
+    output_dir = Path(output_dir)
+    output_dir.mkdir(parents=True, exist_ok=True)
+    output_dir = str(output_dir)
+    callback = partial(trace_handler, output_dir=output_dir)
+
+    profiler = torch.profiler.profile(
+        activities=activities,
+        profile_memory=profile_memory,
+        with_stack=with_stack,
+        record_shapes=record_shapes,
+        with_flops=with_flops,
+        schedule=schedule,
+        experimental_config=experimental_config,
+        on_trace_ready=callback,
+    )
+    if return_cfg:
```

Comment:
How about just return config by default and let recipe decide if it wants to log it?

Reply:
Done